### PR TITLE
fix: prevent traversal path during sos download

### PR DESCRIPTION
**Pull request by quentinalbertone**
# Description
Prevent path traversal issue during sos download.

## Checklist
(For exoscale contributors)

* [x] Changelog updated (under *Unreleased* block, and add the Pull Request #number for each bit you add to the `CHANGELOG.md`)
* [x] Testing

## Testing
```Bash
$> exo storage upload a.txt sos://my-bucket-123456/../a.txt
$> exo storage upload b.txt sos://my-bucket-123456/b.txt
$> mkdir /tmp/test
$> go run main.go storage download -r sos://my-bucket-123456 /tmp/test/ 
warning: Skipping file ../a.txt. File references a parent directory
b.txt     [==============================================================================] 3.13 KiB / 3.13 KiB | 0s
```

### CODE REVIEW

**Comment by quentinalbertone** [comment]
for information
```Bash
$> aws s3 cp s3://my-bucket-123456 test --recursive
warning: Skipping file ../a.txt. File references a parent directory.
download: s3://my-bucket-123456/b.txt to test/b.txt
```

File: pkg/storage/sos/object_test.go [hunk 1]
```diff
@@ -578,3 +578,33 @@ func TestUploadFiles(t *testing.T) {
...
+			expect: true,
+		},
+		{
+			path:   "a/b/../../../test.text",
+			expect: true,
+		},
+		{
+			path:   "a/b/../../test.txt",
+			expect: false,
+		},
```
Comment (natalie-o-perret) [inline pkg/storage/sos/object_test.go:602]: @quentinalbertone I think you might wanna try this edge-case below:

```go
{
	path:   "../../a/b/c/d/e",
	expect: true,
}
```

Output:
```
 go test ./pkg/storage/sos/... -run Test
_IsTraversalPath -v 2>&1
=== RUN   Test_IsTraversalPath
=== RUN   Test_IsTraversalPath/test.txt
=== RUN   Test_IsTraversalPath/../test.txt
=== RUN   Test_IsTraversalPath/a/b/../../../test.text
=== RUN   Test_IsTraversalPath/a/b/../../test.txt
=== RUN   Test_IsTraversalPath/../../a/b/c/d/e
    object_test.go:613: 
                Error Trace:    /home/natalie/Work/Repositories/giza/public-tooling/cli/pkg/storage/sos/object_test.go:613
                Error:          Not equal: 
                                expected: true
                                actual  : false
                Test:           Test_IsTraversalPath/../../a/b/c/d/e
--- FAIL: Test_IsTraversalPath (0.00s)
    --- PASS: Test_IsTraversalPath/test.txt (0.00s)
    --- PASS: Test_IsTraversalPath/../test.txt (0.00s)
    --- PASS: Test_IsTraversalPath/a/b/../../../test.text (0.00s)
    --- PASS: Test_IsTraversalPath/a/b/../../test.txt (0.00s)
    --- FAIL: Test_IsTraversalPath/../../a/b/c/d/e (0.00s)
FAIL
FAIL    github.com/exoscale/cli/pkg/storage/sos 0.005s
?       github.com/exoscale/cli/pkg/storage/sos/object  [no test files]
FAIL
```

File: pkg/storage/sos/object.go [hunk 2]
```diff
@@ -806,3 +806,18 @@ func (o *ShowObjectOutput) ToTable() {
 		return buf.String()
 	}()})
 }
+
+func IsTraversalPath(key string) bool {
```
Comment (natalie-o-perret) [inline pkg/storage/sos/object.go:810]: The algorithm tracks a depth starting at `-1` and returns `true` only if the final value is `< 0`. 

But it only checks the end state, not whether the path escapes at any point.

E.g., `../../a/b/c/d/e`:

| step  | depth |
| ----- | ----- |
| start | -1    |
| `..`  | -2    |
| `..`  | -3    |
| `a`   | -2    |
| `b`   | -1    |
| `c`   | 0     |
| `d`   | 1     |
| `e`   | 2     |

This would returns `false`, **not flagged as traversal**, but `dest/../../a/b/c/d/e` resolves outside the target directory. 

As it stands, it looks like a bypass to me.

I think a more secure implementation would also track the minimum depth, not just the final, wdyt?

[Edit]

Using `path.Clean` like below could also be more go-idiomatic while covering the bypass case, I've mentioned above:
```go
func IsTraversalPath(key string) bool {
    cleaned := path.Clean(key)
    return cleaned == ".." || strings.HasPrefix(cleaned, "../")
}
```

**Review by natalie-o-perret — CHANGES_REQUESTED** [review]
Nice tackling this, but you might need to double check the bypass case I've pointed out.

File: pkg/storage/sos/object.go [hunk 3]
```diff
@@ -806,3 +806,18 @@ func (o *ShowObjectOutput) ToTable() {
...
+	path := strings.Split(key, "/")
```
Comment (kobajagi) [inline pkg/storage/sos/object.go]:
```suggestion
	path := strings.Split(key, filepath.Separator)
```
Windows enters the chat 😉

**Review by natalie-o-perret — APPROVED** [review]

**Review by kobajagi — APPROVED** [review]
